File: scratch/find_missing_response.mjs
import fs from "node:fs";
import path from "node:path";

const baseDir = "c:\\Users\\aou\\Desktop\\Alice\\Study\\side project\\restaurant map";
const aiReviewDir = path.join(baseDir, "ai_review");
const responseDir = path.join(baseDir, "response");

// 1. Read files in ai_review
const aiFiles = fs.readdirSync(aiReviewDir).filter(f => f.endsWith(".json"));
const skippedIds = [];

for (const file of aiFiles) {
    const placeId = path.basename(file, ".json");
    const responsePath = path.join(responseDir, `${placeId}.json`);
    if (!fs.existsSync(responsePath)) {
        skippedIds.push(placeId);
    }
}

console.log(`Total skipped place IDs (missing response JSON): ${skippedIds.length}`);

// 2. Build a mapping of place_id -> name
const idToName = new Map();

// Parse root index.js using regex
const indexPath = path.join(baseDir, "index.js");
if (fs.existsSync(indexPath)) {
    try {
        const indexContent = fs.readFileSync(indexPath, "utf8");
        // Regex to match "place_id": "..." and "name": "..." in close proximity
        const recordRegex = /"place_id":\s*"([^"]+)"[\s\S]*?"name":\s*"([^"]+)"/g;
        let match;
        while ((match = recordRegex.exec(indexContent)) !== null) {
            const pid = match[1].trim();
            const name = match[2].trim();
            idToName.set(pid, name);
        }
        console.log(`Parsed ${idToName.size} place_id -> name mappings from index.js`);
    } catch (err) {
        console.error("Error reading/parsing index.js:", err.message);
    }
}

// Simple CSV parser helper
function parseCsvLine(line) {
    const result = [];
    let current = "";
    let inQuotes = false;
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            result.push(current);
            current = "";
        } else {
            current += char;
        }
    }
    result.push(current);
    return result;
}

// Parse aggregated_restaurants.csv
const aggregatedPath = path.join(baseDir, "aggregated_restaurants.csv");
if (fs.existsSync(aggregatedPath)) {
    try {
        const content = fs.readFileSync(aggregatedPath, "utf8").replace(/^\uFEFF/, "");
        const lines = content.split(/\r?\n/);
        if (lines.length > 0) {
            const header = parseCsvLine(lines[0]);
            const nameIdx = header.indexOf("餐廳名稱");
            const urlIdx = header.indexOf("Google Map網址");
            
            if (nameIdx !== -1 && urlIdx !== -1) {
                let count = 0;
                for (let i = 1; i < lines.length; i++) {
                    if (!lines[i].trim()) continue;
                    const row = parseCsvLine(lines[i]);
                    const name = row[nameIdx];
                    const url = row[urlIdx];
                    if (name && url) {
                        const match = url.match(/query_place_id=([^&\s"]+)/);
                        if (match) {
                            const pid = match[1].trim();
                            idToName.set(pid, name.trim());
                            count++;
                        }
                    }
                }
                console.log(`Parsed ${count} place_id -> name mappings from aggregated_restaurants.csv`);
            }
        }
    } catch (err) {
        console.error("Error parsing aggregated_restaurants.csv:", err.message);
    }
}

// Parse restaurants_refined_v8.csv
const refinedPath = path.join(baseDir, "restaurants_refined_v8.csv");
if (fs.existsSync(refinedPath)) {
    try {
        const content = fs.readFileSync(refinedPath, "utf8").replace(/^\uFEFF/, "");
        const lines = content.split(/\r?\n/);
        if (lines.length > 0) {
            const header = parseCsvLine(lines[0]);
            const idIdx = header.indexOf("place_id");
            const nameIdx = header.indexOf("name");
            
            if (idIdx !== -1 && nameIdx !== -1) {
                let count = 0;
                for (let i = 1; i < lines.length; i++) {
                    if (!lines[i].trim()) continue;
                    const row = parseCsvLine(lines[i]);
                    const pid = row[idIdx];
                    const name = row[nameIdx];
                    if (pid && name) {
                        idToName.set(pid.trim(), name.trim());
                        count++;
                    }
                }
                console.log(`Parsed ${count} place_id -> name mappings from restaurants_refined_v8.csv`);
            }
        }
    } catch (err) {
        console.error("Error parsing restaurants_refined_v8.csv:", err.message);
    }
}

// 3. Compile results for skipped place IDs
const results = [];
for (const pid of skippedIds) {
    const name = idToName.get(pid) || "Unknown Name (未能從現有資料比對出店名)";
    results.push({ placeId: pid, name: name });
}

// Sort by name
results.sort((a, b) => a.name.localeCompare(b.name, "zh-Hant"));

console.log("\n--- List of Restaurants Missing Response Data ---");
results.forEach((item, index) => {
    console.log(`${index + 1}. ${item.name} (${item.placeId})`);
});

// Write to scratch folder
const outPath = path.join(baseDir, "scratch", "skipped_restaurants_list.txt");
const fileContent = "List of Restaurants with missing response data:\n" +
    results.map((item, index) => `${index + 1}. ${item.name} (Place ID: ${item.placeId})`).join("\n");
fs.writeFileSync(outPath, fileContent, "utf8");
console.log(`\nWritten details to ${outPath}`);

File: scratch/find_missing_response.py
import os
import json
import csv

base_dir = r"c:\Users\aou\Desktop\Alice\Study\side project\restaurant map"
ai_review_dir = os.path.join(base_dir, "ai_review")
response_dir = os.path.join(base_dir, "response")

# 1. Get all place_ids from ai_review
ai_files = [f for f in os.listdir(ai_review_dir) if f.endswith(".json")]
skipped_ids = []

for f in ai_files:
    place_id = os.path.splitext(f)[0]
    response_path = os.path.join(response_dir, f"{place_id}.json")
    if not os.path.exists(response_path):
        skipped_ids.append(place_id)

print(f"Total skipped place IDs (missing response JSON): {len(skipped_ids)}")

# 2. Build a mapping of place_id -> name from all potential files in the workspace
id_to_name = {}

# Helper to add from CSV
def add_from_csv(file_path, id_col, name_col):
    full_path = os.path.join(base_dir, file_path)
    if os.path.exists(full_path):
        try:
            with open(full_path, "r", encoding="utf-8-sig") as f:
                reader = csv.DictReader(f)
                for row in reader:
                    pid = row.get(id_col)
                    name = row.get(name_col)
                    if pid and name:
                        id_to_name[pid.strip()] = name.strip()
        except Exception as e:
            print(f"Error reading CSV {file_path}: {e}")

# Helper to add from JSON list/dict
def add_from_json(file_path):
    full_path = os.path.join(base_dir, file_path)
    if os.path.exists(full_path):
        try:
            with open(full_path, "r", encoding="utf-8-sig") as f:
                data = json.load(f)
                if isinstance(data, list):
                    for item in data:
                        if isinstance(item, dict):
                            pid = item.get("place_id") or item.get("placeId")
                            name = item.get("name") or item.get("displayName")
                            if pid and name:
                                if isinstance(name, dict):
                                    name = name.get("text")
                                id_to_name[pid] = name
                elif isinstance(data, dict):
                    # check if format is id -> dict
                    for pid, val in data.items():
                        if isinstance(val, dict):
                            name = val.get("name") or val.get("displayName")
                            if isinstance(name, dict):
                                name = name.get("text")
                            if name:
                                id_to_name[pid] = name
        except Exception as e:
            print(f"Error reading JSON {file_path}: {e}")

# Add from known files
add_from_json("expanded_restaurants.json")
add_from_json("expanded_restaurants_enriched.json")
add_from_json("taipei_locations_complete.json")
add_from_json("missing_restaurants.json")

add_from_csv("aggregated_restaurants.csv", "place_id", "name")
add_from_csv("annotated_v2.csv", "place_id", "name")
add_from_csv("annotated_v3_with_signals.csv", "place_id", "name")
add_from_csv("annotated_v5_signals_final.csv", "place_id", "name")
add_from_csv("restaurants_refined_v8.csv", "place_id", "name")
add_from_csv("feedback.csv", "place_id", "name")

# 3. Match skipped place IDs and report names
results = []
for pid in skipped_ids:
    name = id_to_name.get(pid, "Unknown Name")
    results.append((pid, name))

# Sort by name
results.sort(key=lambda x: x[1])

print("\n--- List of Restaurants Missing Response Data ---")
for idx, (pid, name) in enumerate(results, 1):
    print(f"{idx}. {name} (Place ID: {pid})")

# Write output to a file for easy reading
out_path = os.path.join(base_dir, "scratch", "skipped_restaurants_list.txt")
with open(out_path, "w", encoding="utf-8") as f:
    f.write("List of Restaurants with missing response data:\n")
    for idx, (pid, name) in enumerate(results, 1):
        f.write(f"{idx}. {name} (Place ID: {pid})\n")
print(f"\nWritten details to {out_path}")
